mlir/lib/Analysis/FlatLinearValueConstraints.cpp

```diff
@@ -46,9 +46,15 @@ struct AffineExprFlattener : public SimpleAffineExprFlattener {
   // inequalities.
   IntegerPolyhedron localVarCst;
 
-  AffineExprFlattener(unsigned nDims, unsigned nSymbols)
+  AffineExprFlattener(unsigned nDims, unsigned nSymbols,
+                      bool addConservativeSemiAffineBounds = false)
       : SimpleAffineExprFlattener(nDims, nSymbols),
-        localVarCst(PresburgerSpace::getSetSpace(nDims, nSymbols)) {}
+        localVarCst(PresburgerSpace::getSetSpace(nDims, nSymbols)),
+        addConservativeSemiAffineBounds(addConservativeSemiAffineBounds) {}
+
+  bool hasUnhandledSemiAffineExpressions() const {
+    return unhandledSemiAffineExpressions;
+  }
```

Comment:
the type of approximation you do for semi-affine expressions has no standard form. This should not be part of AffineExprFlattener.

Reply:
I've split this out into shared `AffineExprFlattenerWithLocalVars`, which is subclassed by `AffineExprFlattener` and `SemiAffineExprFlattener`. The `AffineExprFlattener` will fail if it encounters a semi-affine expression, and the `SemiAffineExprFlattener` will attempt to add an approximation.

Reply:
great